ensure tests are compatible with latest pytest

Specifically, fixtures can no longer be directly called. this results in (perhaps among other issues):
```
__________ ERROR collecting allensdk/test/api/test_cell_types_api.py ___________
allensdk/test/api/test_cell_types_api.py:96: in <module>
    @pytest.mark.skipif(cell_types_api() is None, reason='No TEST_API_ENDPOINT set.')
E   _pytest.warning_types.RemovedInPytest4Warning: Fixture "cell_types_api" called directly. Fixtures are not meant to be called directly, are created automatically when test functions request them as parameters. See https://docs.pytest.org/en/latest/fixture.html for more information.
_____________ ERROR collecting allensdk/test/api/test_glif_api.py ______________
allensdk/test/api/test_glif_api.py:63: in <module>
    @pytest.mark.skipif(glif_api() is None, reason='No TEST_API_ENDPOINT set.')
E   _pytest.warning_types.RemovedInPytest4Warning: Fixture "glif_api" called directly. Fixtures are not meant to be called directly, are created automatically when test functions request them as parameters. See https://docs.pytest.org/en/latest/fixture.html for more information.
```

We do want to use the latest pytest, so the scope of work here is to fix these and any other issues that the new version introduces.

--- a/.gitignore
+++ b/.gitignore
@@ -23,3 +23,4 @@ debug.log
 
 # editor configs
 .vscode/
+.mypy_cache/

--- a/allensdk/test/api/test_cell_types_api.py
+++ b/allensdk/test/api/test_cell_types_api.py
@@ -93,7 +93,8 @@ def cell_types_api():
     else:
         return None
 
-@pytest.mark.skipif(cell_types_api() is None, reason='No TEST_API_ENDPOINT set.')
+
+@pytest.mark.requires_api_endpoint
 def test_list_cells_unmocked(cell_types_api):
     from allensdk.config import enable_console_log
     enable_console_log()

--- a/allensdk/test/api/test_glif_api.py
+++ b/allensdk/test/api/test_glif_api.py
@@ -60,7 +60,7 @@ def glif_api():
         return None
 
 
-@pytest.mark.skipif(glif_api() is None, reason='No TEST_API_ENDPOINT set.')
+@pytest.mark.requires_api_endpoint
 def test_get_neuronal_model_templates(glif_api):
 
     assert len(glif_api.get_neuronal_model_templates()) == 7
@@ -85,7 +85,7 @@ def test_get_neuronal_model_templates(glif_api):
             raise Exception('Unrecognized template: %s (%s)' % (template['id'], template['name']))
 
 
-@pytest.mark.skipif(glif_api() is None, reason='No TEST_API_ENDPOINT set.')
+@pytest.mark.requires_api_endpoint
 def test_get_neuronal_models(glif_api, specimen_id):
 
     cells = glif_api.get_neuronal_models([specimen_id])
@@ -93,13 +93,13 @@ def test_get_neuronal_models(glif_api, specimen_id):
     assert len(cells) == 1
     assert len(cells[0]['neuronal_models']) == 2
 
-@pytest.mark.skipif(glif_api() is None, reason='No TEST_API_ENDPOINT set.')
+@pytest.mark.requires_api_endpoint
 def test_get_neuronal_models_no_ids(glif_api):
     cells = glif_api.get_neuronal_models()
     assert len(cells) > 0
 
 
-@pytest.mark.skipif(glif_api() is None, reason='No TEST_API_ENDPOINT set.')
+@pytest.mark.requires_api_endpoint
 def test_get_neuron_configs(glif_api, specimen_id):
     model = glif_api.get_neuronal_models([specimen_id])
 
@@ -110,7 +110,7 @@ def test_get_neuron_configs(glif_api, specimen_id):
 
     np.testing.assert_almost_equal(glif_api.get_neuron_configs([test_id])[test_id]['th_inf'], 0.024561992461740227)
 
-@pytest.mark.skipif(glif_api() is None, reason='No TEST_API_ENDPOINT set.')
+@pytest.mark.requires_api_endpoint
 def test_deprecated(fn_temp_dir, glif_api, neuronal_model_id):
 
     # Exercising deprecated functionality

--- a/allensdk/test/brain_observatory/test_notebook.py
+++ b/allensdk/test/brain_observatory/test_notebook.py
@@ -52,8 +52,8 @@ def boc(tmpdir_factory):
     endpoint = os.environ['TEST_API_ENDPOINT'] if 'TEST_API_ENDPOINT' in os.environ else 'http://twarehouse-backup'
     return BrainObservatoryCache(manifest_file=str(manifest_file), base_uri=endpoint)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+
+@pytest.mark.nightly
 def test_brain_observatory_trace_analysis_notebook(boc):
     # Drifting Gratings
     data_set = boc.get_ophys_experiment_data(502376461)
@@ -92,8 +92,7 @@ def test_brain_observatory_trace_analysis_notebook(boc):
     trial_timestamps = np.arange(-1*dg.interlength, dg.interlength+dg.sweeplength, 1.)/dg.acquisition_rate
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_brain_observatory_static_gratings_notebook(boc):
     data_set = boc.get_ophys_experiment_data(510938357)
     sg = StaticGratings(data_set)
@@ -103,8 +102,7 @@ def test_brain_observatory_static_gratings_notebook(boc):
     assert np.isclose(peak_head['reliability_sg'][0], -0.010099250163301616)
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_brain_observatory_natural_scenes_notebook(boc):
     data_set = boc.get_ophys_experiment_data(510938357)
     ns = NaturalScenes(data_set)
@@ -113,8 +111,8 @@ def test_brain_observatory_natural_scenes_notebook(boc):
     assert np.isclose(ns_head['peak_dff_ns'][0], 4.9614532738)
     assert ns_head['cell_specimen_id'][0] == 517399188
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+
+@pytest.mark.nightly
 def test_brain_observatory_locally_sparse_noise_notebook(boc):
     specimen_id = 517410165
     cell = boc.get_cell_specimens(ids=[specimen_id])[0]
@@ -132,8 +130,7 @@ def test_brain_observatory_locally_sparse_noise_notebook(boc):
     #assert cell_loc
     #assert receptive_field
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_brain_observatory_experiment_containers_notebook(boc):
     targeted_structures = boc.get_all_targeted_structures()
     visp_ecs = boc.get_experiment_containers(targeted_structures=['VISp'])

--- a/allensdk/test/brain_observatory/test_session_analysis.py
+++ b/allensdk/test/brain_observatory/test_session_analysis.py
@@ -81,8 +81,7 @@ def session_c():
     return sa
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 @pytest.mark.parametrize('plot_flag',[False])
 def test_session_a(session_a, plot_flag):
     with patch('allensdk.core.brain_observatory_nwb_data_set.BrainObservatoryNwbDataSet.get_stimulus_table',
@@ -92,8 +91,7 @@ def test_session_a(session_a, plot_flag):
         assert True
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 @pytest.mark.parametrize('plot_flag',[False])
 def test_session_b(session_b, plot_flag):
     with patch('allensdk.core.brain_observatory_nwb_data_set.BrainObservatoryNwbDataSet.get_stimulus_table',
@@ -103,8 +101,7 @@ def test_session_b(session_b, plot_flag):
         assert True
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial_testing")
+@pytest.mark.nightly
 @pytest.mark.parametrize('plot_flag',[False])
 def test_session_c(session_c, plot_flag):
     with patch('allensdk.core.brain_observatory_nwb_data_set.BrainObservatoryNwbDataSet.get_stimulus_table',
@@ -114,24 +111,21 @@ def test_session_c(session_c, plot_flag):
         assert True
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_session_get_session_type(session_a):
     session_type = session_a.nwb.get_session_type()
 
     assert session_type == 'three_session_A'
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_session_get_session_type_b(session_b):
     session_type = session_b.nwb.get_session_type()
 
     assert session_type == 'three_session_B'
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_session_get_session_type_c(session_c):
     session_type = session_c.nwb.get_session_type()
 

--- a/allensdk/test/brain_observatory/test_session_analysis_regression.py
+++ b/allensdk/test/brain_observatory/test_session_analysis_regression.py
@@ -180,16 +180,14 @@ def compare_peak(p1, p2):
         logging.debug("checking " + col)
         assert all(p1[col] == p2[col])
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_session_a(analysis_a, analysis_a_new):
     peak = pd.read_hdf(analysis_a, "analysis/peak")
     new_peak = pd.read_hdf(analysis_a_new, "analysis/peak")
     compare_peak(peak, new_peak)
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_drifting_gratings(dg, nwb_a, analysis_a_new):
     logging.debug("reading outputs")
     dg_new = DriftingGratings.from_analysis_file(BODS(nwb_a), analysis_a_new)
@@ -201,8 +199,7 @@ def test_drifting_gratings(dg, nwb_a, analysis_a_new):
     assert np.allclose(dg.signal_correlation, dg_new.signal_correlation, equal_nan=True)
     assert np.allclose(dg.representational_similarity, dg_new.representational_similarity, equal_nan=True)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_natural_movie_one_a(nm1a, nwb_a, analysis_a_new):
     nm1a_new = NaturalMovie.from_analysis_file(BODS(nwb_a), analysis_a_new, si.NATURAL_MOVIE_ONE)
     #assert np.allclose(nm1a.sweep_response, nm1a_new.sweep_response)
@@ -211,23 +208,20 @@ def test_natural_movie_one_a(nm1a, nwb_a, analysis_a_new):
     assert np.allclose(nm1a.binned_dx_sp, nm1a_new.binned_dx_sp, equal_nan=True)
     assert np.allclose(nm1a.binned_dx_vis, nm1a_new.binned_dx_vis, equal_nan=True)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")    
+@pytest.mark.nightly    
 def test_natural_movie_three(nm3, nwb_a, analysis_a_new):
     #nm3_new = NaturalMovie.from_analysis_file(BODS(nwb_a), analysis_a_new, si.NATURAL_MOVIE_THREE)
     #assert np.allclose(nm3.sweep_response, nm3_new.sweep_response)
     pass
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_session_b(analysis_b, analysis_b_new):
     peak = pd.read_hdf(analysis_b, "analysis/peak")
     new_peak = pd.read_hdf(analysis_b_new, "analysis/peak")
     compare_peak(peak, new_peak)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_static_gratings(sg, nwb_b, analysis_b_new):
     sg_new = StaticGratings.from_analysis_file(BODS(nwb_b), analysis_b_new)
     #assert np.allclose(sg.sweep_response, sg_new.sweep_response)
@@ -238,8 +232,7 @@ def test_static_gratings(sg, nwb_b, analysis_b_new):
     assert np.allclose(sg.signal_correlation, sg_new.signal_correlation, equal_nan=True)
     assert np.allclose(sg.representational_similarity, sg_new.representational_similarity, equal_nan=True)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_natural_movie_one_b(nm1b, nwb_b, analysis_b_new):    
     nm1b_new = NaturalMovie.from_analysis_file(BODS(nwb_b), analysis_b_new, si.NATURAL_MOVIE_ONE)
     #assert np.allclose(nm1b.sweep_response, nm1b_new.sweep_response)
@@ -249,8 +242,7 @@ def test_natural_movie_one_b(nm1b, nwb_b, analysis_b_new):
     assert np.allclose(nm1b.binned_dx_sp, nm1b_new.binned_dx_sp, equal_nan=True)
     assert np.allclose(nm1b.binned_dx_vis, nm1b_new.binned_dx_vis, equal_nan=True)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_natural_scenes(ns, nwb_b, analysis_b_new):    
     ns_new = NaturalScenes.from_analysis_file(BODS(nwb_b), analysis_b_new)
     #assert np.allclose(ns.sweep_response, ns_new.sweep_response)
@@ -260,15 +252,13 @@ def test_natural_scenes(ns, nwb_b, analysis_b_new):
     assert np.allclose(ns.signal_correlation, ns_new.signal_correlation, equal_nan=True)
     assert np.allclose(ns.representational_similarity, ns_new.representational_similarity, equal_nan=True)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_session_c(analysis_c, analysis_c_new):
     peak = pd.read_hdf(analysis_c, "analysis/peak")
     new_peak = pd.read_hdf(analysis_c_new, "analysis/peak")
     compare_peak(peak, new_peak)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_locally_sparse_noise(lsn, nwb_c, analysis_c_new):
     ds = BODS(nwb_c)
     session_type = ds.get_metadata()['session_type']
@@ -282,8 +272,7 @@ def test_locally_sparse_noise(lsn, nwb_c, analysis_c_new):
     #assert np.allclose(lsn.sweep_response, lsn_new.sweep_response)
     assert np.allclose(lsn.mean_sweep_response, lsn_new.mean_sweep_response, equal_nan=True)
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_natural_movie_one_c(nm1c, nwb_c, analysis_c_new):    
     nm1c_new = NaturalMovie.from_analysis_file(BODS(nwb_c), analysis_c_new, si.NATURAL_MOVIE_ONE)
     #assert np.allclose(nm1c.sweep_response, nm1c_new.sweep_response)

--- a/allensdk/test/core/test_mouse_connectivity_notebook.py
+++ b/allensdk/test/core/test_mouse_connectivity_notebook.py
@@ -38,8 +38,7 @@ import pytest
 import os
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_notebook(tmpdir_factory):
 
     # coding: utf-8

--- a/allensdk/test/core/test_nwb_data_set.py
+++ b/allensdk/test/core/test_nwb_data_set.py
@@ -56,24 +56,21 @@ def data_set(request):
     data_set = NwbDataSet(nwb_file)
     return data_set
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_get_sweep_numbers(data_set):
     sweep_numbers = data_set.get_sweep_numbers()
 
     assert len(sweep_numbers) > 0
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_get_experiment_sweep_numbers(data_set):
     sweep_numbers = data_set.get_experiment_sweep_numbers()
 
     assert len(sweep_numbers) > 0
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_get_spike_times(data_set):
     sweep_numbers = data_set.get_experiment_sweep_numbers()
 
@@ -229,8 +226,7 @@ def test_set_spike_times(mock_data_set):
 
     assert False
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_get_sweep_metadata(data_set):
     sweep_metadata = data_set.get_sweep_metadata(1)
 

--- a/allensdk/test/core/test_reference_space_notebook.py
+++ b/allensdk/test/core/test_reference_space_notebook.py
@@ -38,8 +38,7 @@ import pytest
 import os
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true',
-                    reason="partial testing")
+@pytest.mark.nightly
 def test_notebook(tmpdir_factory):
 
     # coding: utf-8

--- a/allensdk/test/test_inline_examples.py
+++ b/allensdk/test/test_inline_examples.py
@@ -15,7 +15,7 @@ EXAMPLE_DIR = os.path.join(
 EXAMPLES = [filename for filename in os.listdir(EXAMPLE_DIR) if filename.split('.')[-1] == 'py']
 
 
-@pytest.mark.skipif(os.getenv('TEST_COMPLETE') != 'true', reason="partial testing")
+@pytest.mark.nightly
 @pytest.mark.parametrize('script_name', EXAMPLES)
 def test_inline_examples(script_name, tmpdir_factory):
 

--- a/conftest.py
+++ b/conftest.py
@@ -1,3 +1,5 @@
+import os
+
 import matplotlib
 matplotlib.use('agg')
 import pytest  # noqa: E402
@@ -12,3 +14,18 @@ def fn_temp_dir(request):
 @pytest.fixture(scope="module")
 def md_temp_dir(request):
     return temp_dir(request)
+
+
+def pytest_collection_modifyitems(config, items):
+    ''' A pytest magic function. This function is called post-collection and gives us a hook for modifying the 
+    collected items.
+    '''
+
+    skip_api_endpoint_test = pytest.mark.skipif(
+        'TEST_API_ENDPOINT' not in os.environ,
+        reason='this test requires that an API endpoint be specified (set the TEST_API_ENDPOINT environment variable).'
+    )
+
+    for item in items:
+        if 'requires_api_endpoint' in item.keywords:
+            item.add_marker(skip_api_endpoint_test)

--- a/conftest.py
+++ b/conftest.py
@@ -26,6 +26,15 @@ def pytest_collection_modifyitems(config, items):
         reason='this test requires that an API endpoint be specified (set the TEST_API_ENDPOINT environment variable).'
     )
 
+    skip_nightly_test = pytest.mark.skipif(
+        os.getenv('TEST_COMPLETE') != 'true',
+        reason='this test is either time/memory/compute expensive or it depends on resources internal to the Allen Institute. '\
+            'Either way, it does\'nt run by default and must be opted into (it does run in our nightly builds).'
+    )
+
     for item in items:
         if 'requires_api_endpoint' in item.keywords:
             item.add_marker(skip_api_endpoint_test)
+
+        if 'nightly' in item.keywords:
+            item.add_marker(skip_nightly_test)

